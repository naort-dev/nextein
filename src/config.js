
import loadEntries from './load-entries'
import BabiliPlugin from 'babili-webpack-plugin'

export default (original) => {
  const {
    webpack: oWebpack,
    exportPathMap: oExportPathMap
  } = original

  return {
    ...original,
    webpack: function (...args) {
      const our = webpack(...args)
      let their

      if (oWebpack) {
        their = oWebpack(...args)
      }

      return {
        ...our,
        ...their
      }
    },
    exportPathMap: async function () {
      const our = await exportPathMap()
      let their
      if (oExportPathMap) {
        their = await oExportPathMap()
      }

      return {
        ...our,
        ...their
      }
    }
  }
}

export const webpack = (config, { dev }) => {
  config.node = {
    fs: 'empty'
  }

  config.plugins = config.plugins.filter(plugin => {
    return plugin.constructor.name !== 'UglifyJsPlugin'
  })

  if (!dev) {
    config.plugins.push(new BabiliPlugin())
  }

  return config
}

export const exportPathMap = async () => {
  const entries = await loadEntries()

  const map = entries
    .reduce((prev, { data }) => {
      const { url, page = 'post', _entry } = data
      const query = { _entry }
      return page ? {
        ...prev,
        [url]: { page: `/${page}`, query }
      } : prev
    }, {})

  return {
    '/': { page: '/' },
    ...map
  }
}
