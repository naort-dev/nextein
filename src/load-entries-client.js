
const rootModuleRelativePath = require('./root-module')
const relativeResolve = rootModuleRelativePath(require)

module.exports = () => {
  const req = require.context(relativeResolve('posts/'), true, /^\.\/.*\.md$/)
  return req
    .keys()
    .map(req)    
}
