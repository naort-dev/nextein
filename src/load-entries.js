/* global __NEXT_DATA__ */
import glob from 'glob'
import { readFileSync, statSync } from 'fs'
import { resolve, basename, extname, relative, dirname, sep } from 'path'
import fm from 'frontmatter'
import fetch from 'unfetch'

const isServer = typeof window === 'undefined'

export default async (path = 'posts') => {
  return (isServer ? fromServer(path) : fromClient(path))
}

const fromServer = async (entriesPath) => {
  const paths = glob.sync(`${entriesPath}/**/*.md`, { root: process.cwd() })

  return paths
    .map(path => readFileSync(path, 'utf-8'))
    .map(fm)
    .map(addEntry(paths))
    .map(addCategory(entriesPath))
    .map(addUrl)
    .map(addDate)
    .filter(({data}) => data.published !== false)
}

const fromClient = async (path) => {
  // in safari the popstate event is fired when user press back and
  // causes the getInitialProps to be called in the SSR version
  // This will pickup the current props and return it as a workaround
  // https://github.com/zeit/next.js/issues/2360
  if (__NEXT_DATA__.nextExport) {
    return __NEXT_DATA__.props.posts
  }
  const resp = await fetch('/_load_entries')
  return resp.json()
}

export const byFileName = async (path) => {
  return isServer ? byFileNameFromServer(path) : byFileNameFromClient(path)
}

const byFileNameFromServer = (path) => {
  return fm(readFileSync(resolve(process.cwd(), path), 'utf-8'))
}

const byFileNameFromClient = async(path) => {
  // in safari the popstate event is fired when user press back and
  // causes the getInitialProps to be called in the SSR version
  // This will pickup the current props and return it as a workaround
  // https://github.com/zeit/next.js/issues/2360
  if (__NEXT_DATA__.nextExport) {
    return __NEXT_DATA__.props.post
  }
  const resp = await fetch(`/_load_entry/${path}`)
  return resp.json()
}

const addEntry = (paths) => (value, idx) => {
  const { data } = value
  return { ...value, data: { ...data, _entry: paths[idx] } }
}

const addCategory = (entriesPath) => (value, idx) => {
  const { data } = value
  return { ...value, data: { ...data, category: createEntryCategory({ entriesPath, ...data }) } }
}

const addUrl = (value, idx) => {
  const { data } = value
  return { ...value, data: { ...data, url: createEntryURL({ ...data }) } }
}

const addDate = (value, idx) => {
  const { data } = value
  return { ...value, data: { ...data, date: createEntryDate({ ...data }) } }
}

const extractName = (path) => basename(path, extname(path))

const createEntryURL = ({ slug, category, _entry, page = 'post' }) => {
  let url = slug
  if (!slug) {
    const name = extractName(_entry)
    url = `/${category ? `${category}/` : ''}${name}`
  }

  return page ? url : undefined
}

const createEntryDate = ({ _entry, date }) => {
  const name = extractName(_entry)
  const DATE_IN_FILE_REGEX = /^(\d{4}-\d{2}-\d{2})-(.+)$/
  const match = name.match(DATE_IN_FILE_REGEX)

  return (date ? new Date(date) : (match) ? new Date(match[1]) : fileCreationDate(_entry)).toJSON()
}

const fileCreationDate = (path) => {
  const { birthtime } = statSync(path)
  return birthtime
}

const createEntryCategory = ({ entriesPath, category, _entry }) => {
  if (category) return category
  const categorySeparator = '/'
  const root = resolve(process.cwd(), entriesPath)
  const post = resolve(process.cwd(), _entry)

  return relative(root, dirname(post)).replace(sep, categorySeparator)
}
