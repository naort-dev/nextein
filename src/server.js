
import http from 'http'
import next from 'next'
import { parse } from 'url'
import { resolve } from 'path'
import route from 'path-match'

import loadEntries, { byFileNameFromServer } from './load-entries'

export default class Server {

  constructor({ dir = '.',  dev = true }) {
    this.app = next({ dev })

    this.wrapAppConfig()
    // this.readEntries()
    // this.defineRoutes()
  }

  wrapAppConfig () {
    const { webpack } = this.app.config

    this.app.config.webpack = (...args) => {
      const original = webpack ? webpack(...args) : null
      const [ config ] = args
      const { node } = config 

      return {
        node: {
          fs: 'empty',
          ...node
        },
        ...original,
        ...config
      } 
    }
    
  }

  defineRoutes() {
    const { entriesMap } = this

    for (const [k, entry] of entriesMap) {
    }

  }

  async readEntries () {
    const entries = await loadEntries('/posts')
    
    const kv = entries.
    map((entry) => {
      const { data } = entry
      const { url } = data
      return [url, entry]
    })

    this.entriesMap = new Map(kv)
  }

  entriesAsJSON() {
    const { entriesMap } = this
    const entries = []
    for (const [k, entry] of entriesMap) {
      entries.push(entry)
    }

    return JSON.stringify(entries)
  }


  handleRequest = (req, res) => {
    const { entriesMap, app } = this
    const parsedUrl = parse(req.url, true)
    const { pathname, query } = parsedUrl

    const matchEntry = route()('/_load_entry/:path')
    const entryParam = matchEntry(pathname)

    if (pathname == '/_load_entries') {
      res.writeHead(200, {"Content-Type": "application/json"})
      return res.end(this.entriesAsJSON())
    }

    if (entryParam) {
      const path = entryParam.path
      
      if (path) {
        const e = byFileNameFromServer(path)

        res.writeHead(200, {"Content-Type": "application/json"})
        return res.end(JSON.stringify(e))
      }
    }
    
    if (entriesMap.has(pathname)) {
      const entry = entriesMap.get(pathname)
      const page = entry.data.page || `post`
      return app.render(req, res, `/${page}`, { _entry: entry.data._entry} )
    }

    app.handleRequest(req, res, parsedUrl)
  }

  async start (port, hostname) {
    await this.readEntries()
    await this.app.prepare()
    this.http = http.createServer(this.handleRequest)
    await new Promise((resolve, reject) => {
      // This code catches EADDRINUSE error if the port is already in use
      this.http.on('error', reject)
      this.http.on('listening', () => resolve())
      this.http.listen(port, hostname)
    })
  }


}