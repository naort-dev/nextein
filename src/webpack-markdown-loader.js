
const fm = require('frontmatter')

module.exports = function (content) {
   
  const cb = this.async()
  const value = fm(content, { safeLoad: true})
  
  value.data._mdPath = this.currentRequest

  cb(null, 'module.exports = ' + JSON.stringify(value) )
}