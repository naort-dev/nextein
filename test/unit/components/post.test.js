jest.mock('../../../src/load-entries')
jest.mock('remark', () => {
  function mockedRemark() { return new impl() }
  mockedRemark.processSync = jest.fn()
  mockedRemark.use = jest.fn(function() { return this })
  class impl {
    use = mockedRemark.use
    processSync = mockedRemark.processSync
  }
  
  return mockedRemark
})

import React from 'react'
import renderer from 'react-test-renderer'

import remark from 'remark'
import { byFileName } from '../../../src/load-entries'
import withPost, { Content } from '../../../src/components/post'

describe('withPost', () => {
  test('exports HOC withPost as default', () => {
    expect(withPost).toBeDefined()
  })

  test('withPosts should add `post` property to getInitialProps', async () => {
    const expected = { data: {}, content: `` }
    const expectedFileName = 'fake'
    const Component = withPost( ({ post }) => (<div>Test</div>) )
    
    byFileName.mockReturnValueOnce(expected)
    const actual = await Component.getInitialProps({query: {_entry: expectedFileName}})
    
    expect(actual.post).toBeDefined()
    expect(actual.post).toEqual(expect.objectContaining(expected))
    expect(byFileName).toHaveBeenCalledWith(expectedFileName)

  })
})

describe('Content', () => {
  test('exports Content', () => {
    expect(Content).toBeDefined()
  })

  test('Content component should render post content', () => {
    const expectedContent = `lorem ipsum`
    
    remark.processSync.mockReturnValueOnce({contents: (<p>{expectedContent}</p>) })

    const comp = renderer.create(<Content content={expectedContent} />)

    expect(remark.processSync).toHaveBeenCalledWith(expectedContent)
    expect(comp.toJSON()).toMatchSnapshot()

  })
})
