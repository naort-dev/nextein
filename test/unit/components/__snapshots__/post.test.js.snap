// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Content component should render post content 1`] = `
<p>
  lorem ipsum
</p>
`;
